pkg/objstore/s3/s3.go

```diff
@@ -44,6 +46,11 @@ const (
 
 	// SSES3 is the name of the SSE-S3 method for objstore encryption.
 	SSES3 = "SSE-S3"
+
+	// SSEConfigKey is the context key to override SSE config. This feature is used by downstream
+	// projects (eg. Cortex) to inject custom SSE config on a per-request basis. Future work or
+	// refactoring can introduce breaking changes as far as the functionality is preserved.
+	SSEConfigKey = ctxKey(0)
```

Comment:
Instead of making this public, I would keep `SSEConfigKey` private, and provide `SetSSEConfigKeyInContext` function, which can then also use a correct `encrypt.ServerSide` type as parameter.

Reply:
Yeah. I would like to hear @bwplotka feedback as well, cause this is like the `WithOptions()` idea which was discarded during the call.

Reply:
BTW, I think your suggestion is better. Just wanna check with Bartek to avoid going back and forth with changes.

Reply:
I think `SetSSEConfigKeyInContext` function and `SSEConfigKey = ctxKey(0)` is the same really, no? I am fine with any as this is matching what we agreed.

Unless we want to spend that time on abstracting and putting in some variadic `opts ...UploadOptions` 🤔 on Upload method? (as https://github.com/thanos-io/thanos/issues/3784#issuecomment-778330460 suggsestion)

Something like `Upload(ctx context.Context, name string, r io.Reader, opts ...UploadOptions)` with  and `WithSSEConfig(...)` so someone can call `Upload(ctx, "yolo", b, objstore.WithSSEConfig)`. As long it's not S3 specific only.

Reply:
> I think `SetSSEConfigKeyInContext` function and `SSEConfigKey = ctxKey(0)` is the same really, no? 

Exposing key relies on client putting the correctly-typed value to the context, and failing in runtime if that’s not the case. Function can have argument of the correct type, and avoid the problem during compilation. Other than that, yes.